src/synth/fluid_synth.c

```diff
@@ -3816,26 +3816,52 @@ fluid_synth_write_float(fluid_synth_t *synth, int len,
     fluid_return_val_if_fail(rout != NULL, FLUID_FAILED);
 
     fluid_rvoice_mixer_set_mix_fx(synth->eventhandler->mixer, 1);
-    l = synth->cur;
     fluid_rvoice_mixer_get_bufs(synth->eventhandler->mixer, &left_in, &right_in);
 
-    for(i = 0, j = loff, k = roff; i < len; i++, l++, j += lincr, k += rincr)
+    cur = synth->cur;
+    j = loff; k = roff;
+
+    do
     {
         /* fill up the buffers as needed */
-        if(l >= synth->curmax)
+        if(cur >= synth->curmax)
         {
-            int blocksleft = (len - i + FLUID_BUFSIZE - 1) / FLUID_BUFSIZE;
+            int blocksleft = (len + FLUID_BUFSIZE - 1) / FLUID_BUFSIZE;
             synth->curmax = FLUID_BUFSIZE * fluid_synth_render_blocks(synth, blocksleft);
             fluid_rvoice_mixer_get_bufs(synth->eventhandler->mixer, &left_in, &right_in);
-
-            l = 0;
+            cur = 0;
         }
 
-        left_out[j] = (float) left_in[0 * FLUID_BUFSIZE * FLUID_MIXER_MAX_BUFFERS_DEFAULT + l];
-        right_out[k] = (float) right_in[0 * FLUID_BUFSIZE * FLUID_MIXER_MAX_BUFFERS_DEFAULT + l];
+        /* calculate amount of available samples */
+        n = synth->curmax - cur;
+
+        /* keep track of emitted samples */
+        if (n > len) n = len;
+        len -= n;
+
+        /* update pointers to current position */
+        left_in  += cur+n;
+        right_in += cur+n;
+
+        /* set final cursor position */
+        cur += n;
+
+        /* reverse index */
+        n = 0 - n;
+
+        do
+        {
+            left_out[j] = (float) left_in[n];
+            right_out[k] = (float) right_in[n];
+
+            j += lincr;
+            k += rincr;
+        }
+        while (++n < 0);
     }
+    while (len);
 
-    synth->cur = l;
+    synth->cur = cur;
 
     time = fluid_utime() - time;
     cpu_load = 0.5 * (fluid_atomic_float_get(&synth->cpu_load) + time * synth->sample_rate / len / 10000.0);
```

Comment:
When control reaches here, `len` will be zero, causing a DIV by zero. Recompile with `cmake -Denable-trap-on-fpe=1` to get a SIGFPE.

Reply:
Argh, I did not notice the use of len into cpu_load, I will fix it.